File: generate_all_suburbs.py
import os
import re

info_file = r'c:\Users\Pau Rodriguez\Antigravity\Trading\Maid At Home (Chat)\suburbs\suburbs_text'
template_file = r'c:\Users\Pau Rodriguez\Antigravity\Trading\Maid At Home (Chat)\suburb_template.html'
output_dir = r'c:\Users\Pau Rodriguez\Antigravity\Trading\Maid At Home (Chat)\suburbs'

if not os.path.exists(output_dir):
    os.makedirs(output_dir)

with open(template_file, 'r', encoding='utf-8') as f:
    template = f.read()

with open(info_file, 'r', encoding='utf-8') as f:
    raw_data = f.read()

sections = raw_data.split('==================================================== SUBURB: ')[1:]

tags_mapping = {
    '{meta_title}': '{{META_TITLE}}', # We should check if the template has these
    '{meta_description}': '{{META_DESCRIPTION}}',
    '{h1}': '{{H1}}',
    '{hero_title}': '{{HERO_TITLE}}',
    '{hero_description}': '{{HERO_DESCRIPTION}}',
    '{standard_clean_description}': '{{STANDARD_DESCRIPTION}}',
    '{deep_clean_description}': '{{DEEP_DESCRIPTION}}',
    '{move_clean_description}': '{{MOVE_DESCRIPTION}}',
    '{steam_clean_description}': '{{STEAM_DESCRIPTION}}',
    '{trust_section_subtitle}': '{{REDEFINING_SUBTITLE}}',
    '{trust_section_title}': '{{REDEFINING_SUBTITLE}}',
    '{testimonial_intro}': '{{TESTIMONIAL_INTRO}}',
    '{process_section_title}': '{{HOW_DESC}}',
    '{guarantee_text}': '{{GUARANTEE_TEXT}}'
}

highlight_color = 'rgb(165, 180, 252)'

for section in sections:
    lines = [line for line in section.split('\n') if line.strip() != '']
    suburb_name = lines[0].split(' =')[0].strip()
    
    print(f"Generating page for: {suburb_name}...")
    
    page_html = template
    
    # Update paths
    page_html = re.sub(r'(href|src)="(\./)?assets/', r'\1="../assets/', page_html)
    def repl_url(match):
        return f'url({match.group(1)}../assets/{match.group(3)}'
    page_html = re.sub(r'url\(([\'"]?)(\./)?assets/([^)]+)', repl_url, page_html)

    root_links = ['index.html', 'about.html', 'pricing.html', 'contact.html', 'faqs.html', 'booking.html', 'terms&conditions.html']
    for link in root_links:
        page_html = re.sub(rf'href="(\./)?{link}"', f'href="../{link}"', page_html)
        
    placeholders = {
        '{{suburb_name}}': f'<span style="color: {highlight_color};">{suburb_name}</span>',
        '{{SUBURB_NAME}}': suburb_name
    }
    
    # Parse tags manually to avoid regex backtrack
    current_tag = None
    current_content = []
    
    parsed_tags = {}
    
    for line in lines[1:]:
        if line.startswith('{') and line.endswith('}'):
            if current_tag:
                parsed_tags[current_tag] = " ".join(current_content).strip()
            current_tag = line.strip()
            current_content = []
        elif line.startswith('TESTIMONIAL CARDS') or line.startswith('-------------------'):
            if current_tag:
                parsed_tags[current_tag] = " ".join(current_content).strip()
                current_tag = None
            continue
        elif line.startswith('Testimonial ') and 'Text:' in line:
            # handle testimonial start
            current_tag = 'testimonial_card'
            current_content = [line]
        elif current_tag == 'testimonial_card' and (line.startswith('Name:') or line.startswith('Suburb:') or line.startswith('Rating:')):
            current_content.append(line)
        elif current_tag == 'testimonial_card' and line.strip() == '':
            pass
        elif current_tag == 'testimonial_card':
            current_content[-1] += f" {line.strip()}"
        elif current_tag:
            current_content.append(line.strip())
            
    if current_tag and current_tag != 'testimonial_card':
        parsed_tags[current_tag] = " ".join(current_content).strip()

    # Process generic tags
    for tag_key, placeholder in tags_mapping.items():
        if tag_key in parsed_tags:
            content = parsed_tags[tag_key]
            # Wrap suburb name in blue span, only in titles
            if tag_key in ['{h1}', '{hero_title}']:
                content = re.sub(rf'(\b\w+\s+{re.escape(suburb_name)}|{re.escape(suburb_name)})', f'<span style="color: {highlight_color};">\\1</span>', content, flags=re.IGNORECASE)
            placeholders[placeholder] = content

    # Testimonials are trickier - Parse from the overall section
    t_matches = re.finditer(r'Testimonial (\d+) Text:\s*(.*?)\s*Name:\s*(.*?)\s*Suburb:\s*(.*?)\s*Rating:\s*(\d+)', section)
    for t_match in t_matches:
        t_num = t_match.group(1)
        quote = t_match.group(2).strip()
        name = t_match.group(3).strip()
        suburb = t_match.group(4).strip()
        
        quote = re.sub(r'[\n\r]+', ' ', quote)
        if not quote.startswith('"') and not quote.startswith('“'): quote = f'"{quote}'
        if not quote.endswith('"') and not quote.endswith('”'): quote = f'{quote}"'
        
        placeholders[f'{{{{T_CARD_{t_num}_QUOTE}}}}'] = quote
        placeholders[f'{{{{T_CARD_{t_num}_NAME}}}}'] = name
        
        placeholders[f'{{{{T_CARD_{t_num}_SUBURB}}}}'] = suburb

    # Replace all placeholders in HTML
    for key, value in placeholders.items():
        page_html = page_html.replace(key, value)
        
    # Write output
    filename = suburb_name.lower().replace(' ', '-') + '.html'
    out_path = os.path.join(output_dir, filename)
    with open(out_path, 'w', encoding='utf-8') as f:
        f.write(page_html)

print("All suburb pages generated successfully!")

File: generate_all_suburbs.js

const fs = require('fs');
const path = require('path');

const infoFilePath = path.join(__dirname, 'suburbs', 'suburbs_text');
const templatePath = path.join(__dirname, 'suburb_template.html');
const outputDir = path.join(__dirname, 'suburbs');

if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir);
}

function generatePages() {
    const rawData = fs.readFileSync(infoFilePath, 'utf8').replace(/\r\n/g, '\n');
    const template = fs.readFileSync(templatePath, 'utf8');

    // Split by suburb header
    const sections = rawData.split(/==================================================== SUBURB: /);

    for (let i = 1; i < sections.length; i++) {
        let section = sections[i];
        const suburbNameRow = section.split('\n')[0].split(' =')[0].trim();
        const suburbName = suburbNameRow;

        console.log(`Generating page for: ${suburbName}...`);

        let pageHtml = template;

        // Path adjustments for sub-directory
        // Assets in HTML
        pageHtml = pageHtml.replace(/(href|src)="(\.\/)?assets\//g, '$1="../assets/');
        // Assets in CSS url()
        pageHtml = pageHtml.replace(/url\(['"]?(\.\/)?assets\//g, (match) => {
            return match.includes('"') ? 'url("../assets/' : (match.includes("'") ? "url('../assets/" : "url(../assets/");
        });

        // Navigation links
        const rootLinks = ['index.html', 'about.html', 'pricing.html', 'contact.html', 'faqs.html', 'booking.html', 'terms&conditions.html'];
        rootLinks.forEach(link => {
            const re = new RegExp(`href="(\\.\\/)?${link}"`, 'g');
            pageHtml = pageHtml.replace(re, `href="../${link}"`);
        });

        const highlightColor = 'rgb(165, 180, 252)';
        const placeholders = {
            '{{suburb_name}}': `<span style="color: ${highlightColor};">${suburbName}</span>`,
            '{{SUBURB_NAME}}': suburbName
        };

        const escapedSuburb = suburbName.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
        const subRegexPattern = `(\\b\\w+\\s+${escapedSuburb}|${escapedSuburb})`;
        const subRegex = new RegExp(subRegexPattern, 'gi');

        // Improved Tag Extraction (Multi-line)
        const tagsMapping = {
            '{hero_title}': '{{HERO_TITLE}}',
            '{hero_description}': '{{HERO_DESCRIPTION}}',
            '{standard_clean_description}': '{{STANDARD_DESCRIPTION}}',
            '{deep_clean_description}': '{{DEEP_DESCRIPTION}}',
            '{move_clean_description}': '{{MOVE_DESCRIPTION}}',
            '{steam_clean_description}': '{{STEAM_DESCRIPTION}}',
            '{trust_section_title}': '{{REDEFINING_SUBTITLE}}',
            '{testimonial_intro}': '{{TESTIMONIAL_INTRO}}',
            '{process_section_title}': '{{HOW_DESC}}',
            '{guarantee_text}': '{{GUARANTEE_TEXT}}'
        };

        // Extract all tags using a non-greedy catch-all until the next brace or section break
        for (const [tag, placeholder] of Object.entries(tagsMapping)) {
            const escapedTag = tag.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
            // Look for the tag and grab everything until the next tag starting with { on a new line
            const tagRegex = new RegExp(`${escapedTag}\\s*([\\s\\S]*?)(?=\\n\\s*\\{|\\n\\s*-------------------|\\n\\s*={5,}|$)`);
            const match = section.match(tagRegex);
            if (match) {
                let content = match[1].replace(/\r?\n/g, ' ').replace(/\s+/g, ' ').trim();

                // ONLY wrap suburb name in blue span if it's the HERO TITLE
                if (placeholder === '{{HERO_TITLE}}') {
                    content = content.replace(subRegex, `<span style="color: ${highlightColor};">$1</span>`);
                }

                placeholders[placeholder] = content;
                console.log(`  - Found ${placeholder}: ${placeholders[placeholder].substring(0, 50)}...`);
            } else {
                console.warn(`  - Warning: Tag ${tag} not found in section for ${suburbName}`);
            }
        }

        // Improved Testimonial Extraction (Multi-line)
        // Extract the whole TESTIMONIAL CARDS block
        const tBlockMatch = section.match(/TESTIMONIAL CARDS[\s\S]*?-------------------([\s\S]*?)(?==|$)/);
        if (tBlockMatch) {
            const tBlock = tBlockMatch[1];
            for (let t = 1; t <= 5; t++) {
                // Match "Testimonial X Text: [QUOTE] Name: [NAME] Suburb: [SUBURB] Rating: [RATING]"
                const tRegex = new RegExp(`Testimonial ${t} Text:\\s*([\\s\\S]*?)\\s*Name:\\s*([\\s\\S]*?)\\s*Suburb:\\s*([\\s\\S]*?)\\s*Rating:\\s*`, 'm');
                const tMatch = tBlock.match(tRegex);
                if (tMatch) {
                    let quote = tMatch[1].replace(/\r?\n/g, ' ').replace(/\s+/g, ' ').trim();
                    // NO highlight in testimonials per user request

                    // Ensure it starts and ends with quotes (either smart or standard)
                    if (!quote.startsWith('“') && !quote.startsWith('"')) quote = `"${quote}`;
                    if (!quote.endsWith('”') && !quote.endsWith('"')) quote = `${quote}"`;

                    placeholders[`{{T_CARD_${t}_QUOTE}}`] = quote;
                    placeholders[`{{T_CARD_${t}_NAME}}`] = tMatch[2].replace(/\r?\n/g, ' ').replace(/\s+/g, ' ').trim();

                    let subVal = tMatch[3].replace(/\r?\n/g, ' ').replace(/\s+/g, ' ').trim();
                    // NO highlight in testimonials per user request
                    placeholders[`{{T_CARD_${t}_SUBURB}}`] = subVal;
                }
            }
        }

        // Perform Replacements
        for (const [key, value] of Object.entries(placeholders)) {
            const escapedKey = key.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
            const re = new RegExp(escapedKey, 'g');
            pageHtml = pageHtml.replace(re, value || '');
        }

        const fileName = suburbName.toLowerCase().replace(/\s+/g, '-') + '.html';
        const outputPath = path.join(outputDir, fileName);
        fs.writeFileSync(outputPath, pageHtml);
    }

    console.log('Success: All suburb pages generated with full multi-line text and fixed assets.');
}

generatePages();
